### DISCUSSION TITLE
Where can I generate a KEY and a SECRET for the Docker Instalation

### DISCUSSION THREAD

**Discussion by RocketRene** [directus/directus#11848] [Archive]
Hi, the Documentation says I should generate a new Key and Secret 

````
# Make sure to change sensitive values (KEY, SECRET, ...) in production

docker run \
  -p 8055:8055 \
  -e KEY=255d861b-5ea1-5996-9aa3-922530ec40b1 \
  -e SECRET=6116487b-cda1-52c2-b5b5-c8022c45e263 \
  directus/directus
````

But I have no Idea where I can do that. Can I use my Password Manager (LastPass) and just generate 2 long Passwords?

Sorry for the dump question.

**Comment by benhaynes** [comment]
CC @erondpowell

**Comment by mattbisme** [comment]
Personally, I just run a `cat /proc/sys/kernel/random/uuid` and use the result; once for `KEY` and again for `SECRET`. I don't know if there is a better way… I think each just needs to be a long random string. 

Related question: what are the `KEY` and `SECRET` used for? I always assumed for salt/hash, but changing them doesn't seem to affect my ability to login.

**Reply by azrikahar** [reply]
> Personally, I just run a `cat /proc/sys/kernel/random/uuid` and use the result; once for `KEY` and again for `SECRET`. I don't know if there is a better way… I think each just needs to be a long random string.

Thanks for sharing! cc @erondpowell 

Alternatively, there's also `openssl rand (-hex/-base64) <number of bytes>`, such as

- `openssl rand -hex 16` (example output: `96c268dcecf92a9b76e8f0b7e113e8df`)

- `openssl rand -base64 24` (example output: `HBy8Ox5rLhluN7hEYwYVOxapPp6ZFpvX`)

> Related question: what are the `KEY` and `SECRET` used for?

The `KEY` is a unique identifier for the current project, which allows things to have a unique identifier to make sure there's no duplication errors eg. when running multiple node processes for the same project in parallel (load balanced). The `SECRET` is used to sign the JWTs used in auth 👍

> I always assumed for salt/hash, but changing them doesn't seem to affect my ability to login.

Changing it will only affect existing (distributed/given) tokens being invalidated, such as old password reset tokens in an email's link. If by salt/hash you mean the ones for passwords themselves, those are generated by Argon2 instead. You can find more info on it over here: https://docs.directus.io/configuration/config-options/#hashing

**Reply by mattbisme** [reply]
Thanks for the information!

**Reply by syahzuan** [reply]
Thank you. Big help!

**Reply by onlinemoose** [reply]
Awesome - THANK YOU. This really ought to be explained in the Directus documentation.

**Reply by reld** [reply]
Brilliant! this helps a lot.
